modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/db/wal/WalCompactionTest.java

```diff
@@ -134,6 +158,74 @@ private RolloverRecord() {
         cleanPersistenceDir();
     }
 
+    /**
+     * Test vefiries that compression process stops on segment with the last successful CP
+     * and doesn't proceed to the end of WAL archive.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testBinaryRecoveryAfterFullCompaction() throws Exception {
+        int minNumberOfSegmentsToCompress = 5;
+
+        walSegments = minNumberOfSegmentsToCompress;
+        fileIoFactory = new CheckpointFailingIoFactory(false);
+
+        final IgniteEx ig = startGrid(0);
+        ig.cluster().state(ACTIVE);
+
+        IgniteCache<Integer, byte[]> cache = ig.cache(CACHE_NAME);
+        loadData(cache, ENTRIES / 2);
+
+        assertTrue(GridTestUtils.waitForCondition(
+                () -> wal(ig).lastArchivedSegment() >= minNumberOfSegmentsToCompress,
+                getTestTimeout()));
+
+        forceCheckpoint(ig);
+
+        loadData(cache, ENTRIES);
+
+        long lastCpIdx = ((FileWALPointer) database(ig).lastCheckpointMarkWalPointer()).index();
+        forceCheckpoint(ig);
+
+        assertTrue(GridTestUtils.waitForCondition(
+                () -> wal(ig).lastArchivedSegment() > lastCpIdx,
+                getTestTimeout()));
+
+        Collection<?> queue = GridTestUtils.getFieldValue(wal(ig), "segmentAware", "segmentCompressStorage", "segmentsToCompress");
+
+        assertTrue(GridTestUtils.waitForCondition(queue::isEmpty, getTestTimeout()));
+
+        fileIoFactory.startFailing();
```

Comment:
Would it be easier to stop the node and delete end marker? Is there a utility code for that?

Reply:
I tried this, it turned out that when we allow a CP to finish successfully it may result in compressor making progress and removing WAL segments that it shouldn't remove.
So I decided it is better to implement a 'fair' scenario with CP failing 'for real'.

Reply:
Ok, didn't think about it